### DISCUSSION TITLE
why does httpx use `client.stream(method, ...)` rather than a `stream` parameter like other library?

### DISCUSSION THREAD

**Discussion by NewUserHa** [encode/httpx#2119] [Ideas]
Examples:
the [requests](https://github.com/psf/requests) uses https://docs.python-requests.org/en/latest/user/quickstart/#raw-response-content.
the [aiohttp](https://github.com/aio-libs/aiohttp) doesn't reads the HTTP content until `await`.

Since [httpx](https://github.com/encode/httpx) is async-able, could [httpx](https://github.com/encode/httpx) not to read the content(probably a doze MB response) until `await`, or merge `client.stream` to ordinary `client.get/post` for convenient like the [requests](https://github.com/psf/requests)?

**Comment by lovelydinosaur** [comment]
Because when you stream a response body you need to make sure that you close the response. Handling that within a context manager is more appropriate to ensure that *always* occurs properly.

To put it into another context, it's similar to us *only* allowing this kind of file opening API...

```python
# Recommended way to open files in Python,
# since it ensures you can't accidentally fail to close it,
# and the context *shows* the scope within which the resource is used.
open("some_file.txt", "r") as some_file:
    ...
```

And *not* allowing this kind of usage...

```python
# Bad practice. Using a context manager is a better construct.
some_file = open("some_file.txt", "r")

# Oops, I better call `some_file.close()` somewhere.
# And I better make sure it's always called even if I get some unexpected exception along the way.
# I guess I could do that with a `try...finally...` block. Wait a minute why didn't I just use a context manager?
```

**Reply by NewUserHa** [reply]
sounds right. But not related to the topic I think.

httpx docs says:
```py
>>> client = httpx.AsyncClient()
>>> async with client.stream('GET', 'https://www.example.com/') as response:
...     async for chunk in response.aiter_bytes():
...         ...
```
So httpx also could use(Since it's already a `AsyncClient()`, it seems not make sense to read all content even >1MB to the memory as the default.):
```py
>>> client = httpx.AsyncClient()
>>> async with client.get('https://www.example.com/', stream=True) as response:
...     async for chunk in response.aiter_bytes(): # or await response.aread()
...         ...
```
because `client.stream(method, ...)` is redundancy since `get` and `post` is the common cases. there should be at least some shortcuts.

I guess `httpx` should probably more be considered as an `async` client than a `sync` one.(an async version requests probably from https://www.python-httpx.org/compatibility/)

**Reply by lovelydinosaur** [reply]
> I guess httpx should probably more be considered as an async client than a sync one.

Strong disagree. The API consideration here is the same in both cases.

**Reply by NewUserHa** [reply]
for that sentence:
I mean, users who need a simple sync client could use just `requests`. so async probably is an advantage for `httpx`.
Therefore async within `httpx` could be good use cases for new users and a good alternative async client.

**Reply by euri10** [reply]
> open("some_file.txt", "r") as some_file:

you probably mean `with open("some_file.txt", "r") as some_file:` ?

**Reply by lovelydinosaur** [reply]
Right understood. Personally I'd rather the other way around tho.
HTTPX is an alternate HTTP client library in the same space as requests. It happens to *also* support async.

**Reply by lovelydinosaur** [reply]
Yes, thanks for the correction @euri10. 😁

**Reply by NewUserHa** [reply]
back to the topic:
probably `httpx` could use `stream=True` as a parameter instead of a `.stream()` method for convenient and more `requests`-like?
